src/Inciter/Carrier.C

```diff
@@ -244,9 +244,57 @@ Carrier::vol()
 }
 
 void
+Carrier::comvol( const std::vector< std::size_t >& gid,
+                 const std::vector< tk::real >& V )
+// *****************************************************************************
+//  Receive nodal volumes on chare-boundaries
+//! \param[in] gid Global mesh node IDs at which we receive volume contributions
+//! \param[in] V Partial sums of nodal volume contributions to chare-boundary
+//!   nodes
+//! \details This function receives contributions to m_vol, which stores the
+//!   nodal volumes. While m_vol stores own contributions, m_volc collects the
+//!   neighbor chare contributions during communication. This way work on m_vol
+//!   and m_volc is overlapped. The two are combined in totalvol().
+// *****************************************************************************
+{
+  Assert( V.size() == gid.size(), "Size mismatch" );
+
+  for (std::size_t i=0; i<gid.size(); ++i) {
+    auto bid = tk::cref_find( m_bid, gid[i] );
+    Assert( bid < m_volc.size(), "Indexing out of bounds" );
+    m_volc[ bid ] += V[i];
+  }
+
+  if (++m_nvol == m_msum.size()) {
+    m_nvol = 0;
+    contribute(
+       CkCallback(CkReductionTarget(Transporter,volcomplete), m_transporter) );
+  }
+}
+
+void
+Carrier::totalvol()
+// *****************************************************************************
+// Sum mesh volumes and contribute own mesh volume to total volume
+// *****************************************************************************
+{
+  // Combine own and communicated contributions of nodal volumes
+  for (const auto& b : m_bid) {
+    auto lid = tk::cref_find( m_lid, b.first );
+    m_vol[ lid ] += m_volc[ b.second ];
+  }
+
+  // Sum mesh volume to host
+  tk::real V = 0.0;
+  for (auto v : m_v) V += v;
+  contribute( sizeof(tk::real), &V, CkReduction::sum_double,
+    CkCallback(CkReductionTarget(Transporter,totalvol), m_transporter) );
+}
+
+void
 Carrier::stat()
```

Comment:
The issue is that it is not immediately obvious, at least to a non-domain expert, what "inciter" generates. If someone were to tell me "Go in and add a counter for the number of cells that change by more than 10% in any given time step" I would have no idea where to look and would need to add a lot of sentinels and do a lot of grepping. Whereas, if I go check the statistics aggregation section and see "Ah, inciterArray prints cell statistics. I should go look at whatever inciterArray is" I am going to spend more time doing meaningful work.

As it stands now, I would have to hope that you are still being paid to work on this project and have time to help me find it. And if I were actually an external collaborator contributing to an open source project, I am not going to even bother to do that.

That is the reason for software engineering. Making code better and faster is nice, but it is more about ensuring that future workers on a project need <whatever> science knowledge, not quinoa knowledge.

But you are in charge and if you think this approach is the one we should take, disregard this. Just understand that the time we save by not fixing this now is paid tenfold during onboarding of new developers and reduced contributions from the open source community.